### PR TITLE
enable transfer of encrypted stream infos from inputstream to kodi

### PR DESCRIPTION
**Pull request by peak3d**
See title...

Next steps: 
- PR mediacodec wich needs this feature (android)
- Develop dekoder in kodi wich decodes encrypted media using widevine implementation in inputstream. (all platforms != android)

### CODE REVIEW

**Comment by FernetMenta** [comment]
all crypto related stuff should be encapsulated into special crypto classes that reside in an extra file. DemuxPacket can carry crypto info via an unique pointer which is a public member. that keep interference with current code to a minimum

File: xbmc/addons/InputStream.cpp [hunk 1]
```diff
@@ -342,6 +343,18 @@ void CInputStream::UpdateStreams()
         demuxStream->ExtraData[j] = stream.m_ExtraData[j];
     }
 
+    if (stream.crypto_key_system != INPUTSTREAM_INFO::CRYPTO_KEY_SYSTEM_NONE
+      && stream.crypto_key_system < INPUTSTREAM_INFO::CRYPTO_KEY_SYSTEM_COUNT)
```
Comment (FernetMenta) [inline xbmc/addons/InputStream.cpp]: I prefer operator && end of previous line

File: xbmc/addons/InputStream.cpp [hunk 2]
```diff
@@ -342,6 +343,18 @@ void CInputStream::UpdateStreams()
...
+    {
+      static const CryptoSessionSystem map[] = {
```
Comment (FernetMenta) [inline xbmc/addons/InputStream.cpp]: braces next line

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h [hunk 3]
```diff
@@ -135,6 +136,8 @@ class CDemuxStream
   , FLAG_HEARING_IMPAIRED = 0x0080
   , FLAG_VISUAL_IMPAIRED  = 0x0100
   } flags;
+
+  std::shared_ptr<DemuxCryptoSession> crypto_session;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h]: style is always CamelCase

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h [hunk 4]
```diff
@@ -38,4 +40,6 @@ typedef struct DemuxPacket
   double duration; // duration in DVD_TIME_BASE if available
 
   int dispTime;
+
+  std::shared_ptr<DemuxCryptoInfo> cryptoInfo;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h]: DemuxPacket is a struct and we do things like memset on it. having a shared ptr in this struct is probably not a good idea

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxUtils.h [hunk 5]
```diff
@@ -26,6 +26,6 @@ class CDVDDemuxUtils
 {
 public:
   static void FreeDemuxPacket(DemuxPacket* pPacket);
-  static DemuxPacket* AllocateDemuxPacket(int iDataSize = 0);
+  static DemuxPacket* AllocateDemuxPacket(int iDataSize = 0, unsigned int encryptedSubsampleCount = 0);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxUtils.h]: why don't you create a new ctor, setting defaults is error prone, I stopped doing this

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h (hunk 4 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h]: scratch this, we do memset only on the data

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxUtils.h (hunk 5 above)
Comment (peak3d) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxUtils.h]: ok

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h (hunk 4 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h]: maybe cross post. disregard me comment here.

**Comment by peak3d** [comment]
I have changed several places to camelCase and fixed the other issues

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/kodi_inputstream_types.h [hunk 6]
```diff
@@ -46,11 +46,13 @@ extern "C" {
    */
   typedef struct INPUTSTREAM_CAPABILITIES
   {
-    bool m_supportsIDemux;                  /*!< @brief supports interface IDemux */
```
Comment (FernetMenta) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/kodi_inputstream_types.h]: this seems unrelated to the PR.

**Comment by FernetMenta** [comment]
apart from the unrelated change of INPUTSTREAM_CAPABILITIES looks ok now

**Comment by peak3d** [comment]
I'll create an PR for the CAPS, they are now removed from this PR

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h (hunk 4 above)
Comment (AlwinEsch) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h]: This line generate errors on Linux.

```
[ 84%] Building CXX object build/cores/VideoPlayer/CMakeFiles/VideoPlayer.dir/DVDMessage.cpp.o
In file included from /home/alwin/Development/kodi-agile/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxUtils.h:23:0,
                 from /home/alwin/Development/kodi-agile/xbmc/cores/VideoPlayer/DVDMessage.cpp:24:
/home/alwin/Development/kodi-agile/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h:44:8: error: ‘shared_ptr’ in namespace ‘std’ does not name a template type
   std::shared_ptr<DemuxCryptoInfo> cryptoInfo;
        ^
build/cores/VideoPlayer/CMakeFiles/VideoPlayer.dir/build.make:158: die Regel für Ziel „build/cores/VideoPlayer/CMakeFiles/VideoPlayer.dir/DVDMessage.cpp.o“ scheiterte
make[2]: *** [build/cores/VideoPlayer/CMakeFiles/VideoPlayer.dir/DVDMessage.cpp.o] Fehler 1
CMakeFiles/Makefile2:9499: die Regel für Ziel „build/cores/VideoPlayer/CMakeFiles/VideoPlayer.dir/all“ scheiterte
make[1]: *** [build/cores/VideoPlayer/CMakeFiles/VideoPlayer.dir/all] Fehler 2
Makefile:138: die Regel für Ziel „all“ scheiterte
make: *** [all] Fehler 2
```

The crazy thing is on several others who have included `./xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h` it works, is VideoPlayer parts the `C++11` not set to use?

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h]: should include <memory> when using shared_ptr

Comment (AlwinEsch) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h]: Has tested with them, does not work. I log now the cmake files.

**Comment by FernetMenta** [comment]
@peak3d please add new header to cmake
